[24.2] Better cleanup of sharing roles on user purge

jdavcs opened the pull request:
Builds on top of #18966 (depends on a relocated fixture from that PR, nothing more)

Currently, a user's email address is used in a sharing role's name. It's not practical to remove it because a sharing role must be identified somehow in the UI; it is possible to pull the current email for each associated user for each sharing role, but that seems like overkill, especially when multiple sharing roles need to be displayed, each potentially associated with multiple users. Using an email in the role name is for role description only: we don't retrieve roles based on email comparison anymore (fixed in #18942). 

However, when a user is purged, we don't remove their email from associated sharing roles' names. This PR fixed this: we replace the user's email in the sharing role name with the string "[USER PURGED]", so a sharing role for john and alice, after john is purged, will appear as "sharing role for [USER PURGED], alice". If this sharing role has only one user association remaining with the user being purged, the role will be deleted. This also ensures there'll be no orphan sharing roles in the db (a sharing role without an associated user).

Test included.

To do in a follow-up: handle user email updates (in `api/users.py`) by updating email in sharing role name.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Cleanup sharing roles on user purge, refactor, test

1. Update sharing role name removing purged user email
2. If sharing role has only one user association, delete it.
3. Factor out cleanup, add test.

Note: this will not work if user email has been udpated after the sharing
role was created.

jdavcs committed:
Move method out of managers into model

To accommodate package unit testing

bgruening commented on lib/galaxy/webapps/galaxy/controllers/admin.py:
```diff
@@ -204,8 +205,10 @@ def apply_query_filter(self, query, **kwargs):
                     q = term.text
                     if key == "name":
-                        query = query.filter(text_column_filter(self.model_class.name, term))
+                        pass
```
Should that be removed?

jdavcs replied:
Removed. Thanks, sorry I missed those.

bgruening commented on lib/galaxy/webapps/galaxy/controllers/admin.py:
```diff
@@ -204,8 +205,10 @@ def apply_query_filter(self, query, **kwargs):
+                        pass
+                        query = query.filter(text_column_filter(self.model_class._name, term))
                     if key == "description":
+                        pass
```
remove?

bgruening commented on lib/galaxy/webapps/galaxy/api/users.py:
```diff
@@ -730,6 +731,19 @@ def send_activation_email(
+    @router.get(
+        "/api/users/{user_id}/roles",
+        name="get user roles",
+        description="Return a collection of roles associated with this user. Only admins can see user roles.",
```
We should avoid the name `collection` here if possible to avoid confusion with "Galaxy's collections".

jdavcs replied:
Replaced with "list". I think list is generic enough in this context, but we can reword if needed.

jdavcs committed:
Cleanup wording, left over debug code

jdavcs committed:
Rebuild schema

arash77 committed:
Enhance filename handling in data download and zipstream headers to support UTF-8 encoding

arash77 committed:
Add tests for datasets and collections download with non english characters

jdavcs committed:
Update version to 24.3.dev0

jdavcs committed:
Merge branch 'release_24.2' into 'dev'

jdavcs committed:
Merge pull request #19171 from jdavcs/version-24.3.dev

Version 24.3.dev

arash77 committed:
Refactor filename handling in data download to use a dedicated toContentDisposition function for improved UTF-8 support

arash77 committed:
Sanitize filenames in toContentDisposition function to ensure valid characters are used

arash77 committed:
Rename toContentDisposition function to to_content_disposition
apply 255 characters limit to filename
exclude `/\?%*:|"<>` characters from encoded filename

arash77 committed:
exclude ext from character_limit

jdavcs committed:
Fix version: 25.0

mvdbeek committed:
Merge pull request #19173 from jdavcs/dev_fix_version

Fix version: 24.3 >> 25.0

mvdbeek committed:
Merge pull request #19161 from arash77/add-utf8-encoded-filename

Enhance UTF-8 support for filename handling in downloads
